galsim/bandpass.py

```diff
@@ -83,90 +83,122 @@ class Bandpass(object):
-                elif blue_limit is not None:
-                    test_wave = blue_limit
-                else:
-                    # If neither `blue_limit` nor `red_limit` is defined, then the Bandpass should
-                    # be able to be evaluated at any wavelength, so check.
-                    test_wave = 700
-                try:
-                    tp = eval('lambda wave : ' + tp)
-                    tp(test_wave)
-                except:
-                    raise ValueError(
-                        "String throughput must either be a valid filename or something that "+
-                        "can eval to a function of wave. Input provided: {0}".format(throughput))
+        self._orig_tp = throughput  # Save this for pickling.
+        self._tp = _tp              # This is orig_tp turned into an actual function
+
+        if blue_limit > red_limit:
+            raise ValueError("blue_limit must be less than red_limit")
+        self.blue_limit = blue_limit # These may change as we go through this.
+        self.red_limit = red_limit
+        self.zeropoint = zeropoint
+        self.wave_type = wave_type
 
         # Figure out wavelength type
-        if wave_type.lower() in ['nm', 'nanometer', 'nanometers']:
-            wave_factor = 1.0
-        elif wave_type.lower() in ['a', 'ang', 'angstrom', 'angstroms']:
-            wave_factor = 10.0
+        if self.wave_type.lower() in ['nm', 'nanometer', 'nanometers']:
+            self.wave_factor = 1.0
+        elif self.wave_type.lower() in ['a', 'ang', 'angstrom', 'angstroms']:
+            self.wave_factor = 10.0
         else:
-            raise ValueError("Unknown wave_type '{0}'".format(wave_type))
+            raise ValueError("Unknown wave_type '{0}'".format(self.wave_type))
+
+        # Convert string input into a real function (possibly a LookupTable)
+        self._initialize_tp()
+
+        if _wave_list is not None:
+            # Manual override!  Be careful!
+            self.wave_list = _wave_list
+            # This also means that red_limit and blue_limit are already set correctly.
+            # Don't change them.
+            assert self.blue_limit is not None
+            assert self.red_limit is not None
+            return
+
+        # Account for wave_factor in wavelength limits
+        if self.wave_factor != 1.0:
+            if self.blue_limit is not None:
+                self.blue_limit /= self.wave_factor
+            if self.red_limit is not None:
+                self.red_limit /= self.wave_factor
 
         # Assign blue and red limits of bandpass
-        if isinstance(tp, galsim.LookupTable):
-            if blue_limit is None:
-                blue_limit = tp.x_min
-            if red_limit is None:
-                red_limit = tp.x_max
+        if isinstance(self._tp, galsim.LookupTable):
+            if self.blue_limit is None:
+                self.blue_limit = float(self._tp.x_min)/self.wave_factor
```

Comment:
(If this wasn't the bug exactly, it was something like this.  One of the unit tests failed when it used Angstroms, and it was something about how the wave_factor was applied that didn't quite work correctly in that case.)